GlobalSnapshots/src/main/java/Node.java

```diff
@@ -1,63 +1,104 @@
+                    savedState = state;                          // Save state
+                    snapColor = Color.RED;                       // Red means a snapshot is in progress
+                    incomingChannels.initializeChan();           // Initialize chan collection
+                    incomingChannels.initializeClosed();         // Initialize closed collection
+                    sendMsgToNeighbors(forwardMessage);          // Forward Marker to all outgoing neighbors
+                }
+                incomingChannels.close(chanId);                  // Upon receive "Marker" from a channel, close that channel
+                if (incomingChannels.allClosed()){               // When received "Marker" from all incoming channels,
+                    snapColor = Color.WHITE;                     // Snapshot is done.  Can reset the snapColor to WHITE
+                }
+            }
+            else if (command == "RESTORE") {
+                if (restoreColor == Color.WHITE && snapColor == Color.WHITE) { // No restore or snapshot is in progress. Begin a restore.
+                    restoreColor = Color.RED;                     // Red means restore is in progress
+                    incomingChannels.initializeClosed();          // Initialize restoreClosed collection
+                    sendMsgToNeighbors(forwardMessage);           // Forward Restore to all outgoing neighbors
+                }
+                incomingChannels.close(chanId);                   // Upon receive "Restore" from a channel, close that channel         
+                if (incomingChannels.allClosed()){                // When received "Restore" from all incoming channels, it is safe to restore
+                    state = savedState;                           // Restore state
+                    incomingChannels.restoreTransitMessages(handle);    // Restore state messages
+                    restoreColor = Color.WHITE;                   // Restore is done.  Can reset the restoreColor to WHITE
+                }
             } 
-            fileWriter.close();
-        } catch (Exception err) {
-            System.out.println(err.toString());
+            else if (command == "Exit") {
+                sendMsgToNeighbors(forwardMessage);               // Forward "Exit" to all outgoing neighbors
+                exit = true;
+            }
+            eventLogger.logMessage(receivedMsg, state);
+        } while (exit == false);
+        eventLogger.close();
+    }
+
+    public void sendMsgToNeighbors(Message message) {
+        for (BlockingQueue<Message> handle : this.outgoingChannels){         // forward message to all outgoing channels
+            try {
+                handle.put(message);
+            } catch (InterruptedException err) {
+                System.out.println("Node" + String.valueOf(this.id) + ": sendMsgToNeighbors: " + err.toString());
+            }
         }
     }
+
+    public void insertProcessingTime(int milliseconds) {
```

Comment:
so processing time just makes the entire node sleep and the queue can start building up. in other words it may work for 4 nodes waiting at 200ms each time a new state is in there so 250ms  is ample to disregard one message.  but what if we had 1000 nodes with a single input/output would the value of 250ms still be sufficient? or do we not need to worry about that case?